voltdrop_direct/index-es5.js

```diff
@@ -0,0 +1,328 @@
+        voltageL2: (((rawBytesArray[3] << 8) + rawBytesArray[4]) >>> 0) / 64.0,
+        voltageL3: (((rawBytesArray[5] << 8) + rawBytesArray[6]) >>> 0) / 64.0,
+        // Phase Power Factors - 1 byte each, 8-bit signed integers as percentage
+        powerFactorL1: rawBytesArray[7] << 24 >> 24,
+        powerFactorL2: rawBytesArray[8] << 24 >> 24,
+        powerFactorL3: rawBytesArray[9] << 24 >> 24,
+        // Capacitor Voltage Scalar - 1 byte
+        // 8-bit unsigned integer representing the capacitor voltage.
+        // (as if the integer range from 0-255 is scaled to between 0.0V and 5.0V)
+        capacitorVoltage: rawBytesArray[10] * capacitorVoltageFactor
+      };
+      if (result.data.capacitorVoltage < 3.4) {
+        result.warnings.push("Low capacitor voltage may reduce transmit interval.");
+      }
+    break;
+    case packetList.VD_DIRECT_AMPERAGE:
+      var currentL1 = (((rawBytesArray[1] << 8) + rawBytesArray[2]) >>> 0) / 16.0;
+      var currentL2 = (((rawBytesArray[3] << 8) + rawBytesArray[4]) >>> 0) / 16.0;
+      var currentL3 = (((rawBytesArray[5] << 8) + rawBytesArray[6]) >>> 0) / 16.0;
+      result.data = {
+        // Average Phase Current - 2 bytes each
+        // 16-bit unsigned integers in network byte order (MSB/BE) with 12 integer and 4 fractional bits
+        currentL1: currentL1,
+        currentL2: currentL2,
+        currentL3: currentL3,
+        // Maximum Phase Current - 1 byte each
+        // 8-bit unsigned integer with 3 integer and 5 fractional bits (expressed as percentage in addition to 100%)
+        maxCurrentL1: (rawBytesArray[7] / 32.0 + 1.0) * currentL1,
+        maxCurrentL2: (rawBytesArray[8] / 32.0 + 1.0) * currentL2,
+        maxCurrentL3: (rawBytesArray[9] / 32.0 + 1.0) * currentL3,
+        // Temperature Scalar
+        // 8-bit unsigned integer representing the temperature.
+        // (as if the integer range from 0-255 is scaled to between -40C and 80C)
+        temperatureCelsius: rawBytesArray[10] * temperatureCelsiusFactor - 40.0
+      };
+    break;
+    case packetList.VD_DIRECT_ACT_ENERGY_CONF: // Intentional fall-through
+    case packetList.VD_DIRECT_ACT_ENERGY_UNCONF:
+      result.data = {
+        // Total Forward Active Energy - 8 bytes
+        // Sum of all phases active energy accumulated in Watt-Hours since last factory reset downlink.
+        // 64-bit unsigned integer in network byte order (MSB/BE)
+        //
+        // NOTE: Due to the limitations of Javascript and JSON this codec only uses 2^53 bits
+        // This still gives an effective range of 102,821 years at 10 Mega-Watts which is more than the
+        // Voltdrop is reasonably capable of measuring.
+        // Javascript cannot handle bit-shifts above 16 without truncation so they have been replaced by
+        // equivalent multipliers. rawBytesArray[1] is above the 53-bit limit and intentionally ignored.
+        activeEnergyAccumulation: (((rawBytesArray[2] & 0x1F) * 281474976710656) +
```

Comment:
NICE

Reply:
That was a fun bunch of testing to figure out why numbers were going screwy because large shifts don't work correctly.